### PR TITLE
Fix ink splashes don't survive key reparenting

### LINKED ISSUE #6751: When InkWell has a GlobalKey and changes position, splash should not stop

### PR DESCRIPTION
**Pull request by YeungKC**
## Description

Reproduction code: https://github.com/flutter/flutter/issues/6751#issuecomment-727621925

Currently on master branches, ink does not reset splashes when the key is changed, this PR has been fixed.

## Related Issues

Fixes #6751

## Tests

I added the following tests:

- ink_well_test.dart
  - Reset splashes by Key

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by dkwingsmt** [comment]
If I understand correctly this PR means to reset and remove the reparented splash, but I thought the intention was to continue it?

**Comment by YeungKC** [comment]
I think it is should remove splashs, because if you change the Reproduction code InkWell to FlatButton, then you will also see the behavior of remove splashs.

**Comment by dkwingsmt** [comment]
@Hixie Can you illustrate what you planned by the original issue?

File: packages/flutter/lib/src/material/ink_well.dart [hunk 1]
```diff
@@ -775,6 +775,11 @@ class _InkResponseState extends State<_InkResponseStateWidget>
   @override
   void didUpdateWidget(_InkResponseStateWidget oldWidget) {
     super.didUpdateWidget(oldWidget);
+    if (widget.key != oldWidget.key) {
```
Comment (Hixie) [inline packages/flutter/lib/src/material/ink_well.dart:778]: this is guaranteed to always be false

if you can ever find a case where this is true, we have a serious bug

**Comment by dkwingsmt** [comment]
Just clarified with Hixie privately. The issue actually means that, if the widget is moved to a different place in the tree while holding the same key, its splash should persist.

**Comment by YeungKC** [comment]
Ok, I got it, I'll try to fix it.

File: packages/flutter/lib/src/material/ink_well.dart (hunk 1 above)
Comment (YeungKC) [reply packages/flutter/lib/src/material/ink_well.dart:778]: You are right, I will revertPR and rethink about https://github.com/flutter/flutter/issues/6751#issuecomment-730028574.

**Comment by YeungKC** [comment]
@dkwingsmt Inevitably, I've modified some of the tests you wrote earlier. Please check it out.

**Comment by dkwingsmt** [comment]
Why is it necessary to add `widgetKey` to `InkFeature`? I would imagine an `InkFeature` should be owned by a render box (`_RenderInkFeatures`), therefore should not be transferred to other render objects after the reparent, and the owner should be able to track its features during the down time with a simple list.

File: packages/flutter/test/material/ink_well_test.dart [hunk 2]
```diff
@@ -1332,4 +1332,78 @@ void main() {
...
+        await tester.tap(find.byType(InkWell));
+      await tester.pump(const Duration(milliseconds: 60));
+
+      expect(
+        tester.renderObject<RenderProxyBox>(find.byType(PhysicalModel)).child,
+        useKey ? (paints..circle()) : isNot(paints..circle()),
+      );
+    }
+
+    await test(useKey: false, onTapChangeWrap: true);
```
Comment (dkwingsmt) [inline packages/flutter/test/material/ink_well_test.dart:1403]: Prefer DAMP over DIY in unit tests when there's a conflict. In other words, write each "variation" as a separate `test`. In order to reuse widget code, you can define a testing widget private to this file. You'll probably want to briefly document its usage and parameters as well.

You might reuse test process code if really necessary, but statement such as `useKey ? (paints..circle()) : isNot(paints..circle()),` is unfriendly to test readers while not saving much at all.

Comment (YeungKC) [reply packages/flutter/test/material/ink_well_test.dart:1403]: Makes sense, and I'll update it.

**Pull request closed without merging**